gymnasium/envs/mujoco/f.py

```diff
@@ -0,0 +1,315 @@
+        # TODO change bounds and types when and if `Box` supports JAX nativly
+        self.action_space = gymnasium.spaces.Box(
+            low=np.array(low_action_bound),
+            high=np.array(high_action_bound),
+            dtype=np.float32,
+        )
+        # self.action_space = gymnasium.spaces.Box(low=low_action_bound, high=high_action_bound, dtype=low_action_bound.dtype)
+        # observation_space: gymnasium.spaces.Box  # set by the sub-class
+
+    def initial(self, rng: jax.random.PRNGKey) -> mjx._src.types.Data:
+        # TODO? find a more performant alternative that does not allocate?
+        mjx_data = mjx.make_data(self.model)
+        qpos, qvel = self._gen_init_state(rng)
+        mjx_data = mjx_data.replace(qpos=qpos, qvel=qvel)
+        mjx_data = mjx.forward(self.mjx_model, mjx_data)
+
+        return mjx_data
+
+    def transition(
+        self, state: mjx._src.types.Data, action: jnp.ndarray, rng=None
+    ) -> mjx._src.types.Data:
+        """Step through the simulator using `action` for `self.dt`."""
+        mjx_data = state
+        mjx_data = mjx_data.replace(ctrl=action)
+        mjx_data = jax.lax.fori_loop(
+            0, self.frame_skip, lambda _, x: mjx.step(self.mjx_model, x), mjx_data
+        )
+
+        return mjx_data
+        # TODO fix sensors with MJX>=3.1
+
+    def reward(
+        self,
+        state: mjx._src.types.Data,
+        action: jnp.ndarray,
+        next_state: mjx._src.types.Data,
+    ) -> jnp.ndarray:
+        return self._get_reward(state, action, next_state)[0]
+
+    def transition_info(
+        self,
+        state: mjx._src.types.Data,
+        action: jnp.ndarray,
+        next_state: mjx._src.types.Data,
+    ) -> dict:
+        return self._get_reward(state, action, next_state)[1]
+
+    def render_image(
+        self, state: mjx._src.types.Data, render_state: MujocoRenderer
+    ) -> tuple[MujocoRenderer, np.ndarray | None]:
+        # NOTE function can not be jitted
+        mjx_data = state
+        mujoco_renderer = render_state
+
+        data = mujoco.MjData(self.model)
+        mjx.device_get_into(data, mjx_data)  # TODO use get_data instead once mjx==3.1
+        mujoco.mj_forward(self.model, data)
+
+        mujoco_renderer.data = data
+
+        frame = mujoco_renderer.render(
+            self.render_mode, self.camera_id, self.camera_name
+        )
+
+        return mujoco_renderer, frame
+
+    def render_init(
+        self,
+        default_camera_config: dict[str, float] = {},
+        camera_id: int | None = None,
+        camera_name: str | None = None,
+        max_geom=1000,
+        width=480,
+        height=480,
+        render_mode="rgb_array",
+    ) -> MujocoRenderer:
+        # TODO storing to much state? it should probably be moved internal to MujocoRenderer
+        self.render_mode = render_mode
+        self.camera_id = camera_id
+        self.camera_name = camera_name
+
+        return MujocoRenderer(
```

Comment:
Is that jax native too?

Reply:
No it is not, and the render functions will not be JITable for a while (if ever), because of the structure of the `deepming/mujoco` project